pgzero/storage.py

```diff
@@ -0,0 +1,114 @@
+import json
+import os
+import platform
+import logging
+
+
+logger = logging.getLogger(__name__)
+
+
+__all__ = ['StorageCorruptionException', 'JSONEncodingException', 'Storage']
+
+
+class StorageCorruptionException(Exception):
+    """The data in the storage is corrupted."""
+
+
+class JSONEncodingException(Exception):
+    """The data in the storage is corrupted."""
+
+
+class Storage(dict):
+
+    path = ''
+
+    """Behaves like a dictionary with a few extra functions.
+
+    It's possible to load/save the data to disk.
+
+    The name of the file will be the script's name hashed.
+
+    NOTE: If two scripts have the same name, they will load/save from/to
+    the same file.
+    """
+    def __init__(self):
+        dict.__init__(self)
+
+        storage_path = self._get_platform_pgzero_path()
+        if not os.path.exists(storage_path):
+            os.makedirs(storage_path)
+
+    @classmethod
+    def set_app_hash(cls, name):
+        storage_path = cls._get_platform_pgzero_path()
+        cls.path = os.path.join(storage_path, '{}.json'.format(name))
+
+    def load(self):
+        """Load data from disk."""
+        self.clear()
+
+        try:
+            with open(self.path, 'r') as f:
+                data = json.load(f)
+        except FileNotFoundError:
+            # If no file exists, it's fine
+            logger.debug('No file to load data from.')
+        except json.JSONDecodeError:
+            msg = 'Storage is corrupted. Couldn\'t load the data.'
+            logger.error(msg)
+            raise StorageCorruptionException(msg)
+        else:
+            self.update(data)
+
+    def save(self):
+        """Save data to disk."""
+        try:
+            data = json.dumps(self)
+        except TypeError:
+            json_path, key_type_obj = self._get_json_error_keys(self)
+            msg = 'The following entry couldn\'t be JSON serialized: storage{} of type {}'
+            msg = msg.format(json_path, key_type_obj)
+            logger.error(msg)
+            raise JSONEncodingException(msg)
+        else:
+            with open(self.path, 'w+') as f:
+                f.write(data)
+
+    @classmethod
+    def _get_json_error_keys(cls, obj, json_path=''):
+        """Work out which part of the storage failed to be JSON encoded.
+
+        This function returns None if there is no error.
+        """
+
+        if type(obj) in (float, int, str, bool):
+            return None
+        elif isinstance(obj, list):
+            for i, value in enumerate(obj):
+                result = cls._get_json_error_keys(value, '{}[{}]'.format(json_path, i))
+                if result is not None:
+                    return result
+        elif isinstance(obj, dict):
+            for k, v in obj.items():
+                result = cls._get_json_error_keys(v, '{}[\'{}\']'.format(json_path, k))
```

Comment:
You should check that k is a string: JSON only supports strings as dict keys.

Also on this line, you're better to write:

```python
'{}[{!r}]'.format(json_path, k)
```

as the repr of a string containing special characters is more complicated than `"'{}'".format(s)`

Reply:
This is done, need to write a unittest for it.